### PR TITLE
First pass at a helm chart for router

### PR DESCRIPTION
**Pull request by damienpontifex**
- Mostly defaults from `helm create`
- Supports managed federation with values for `managedFederation.apiKey` and `managedFederation.graphRef`
- Allows setting router configuration in values for helm chart under the `configuration` key in values
- Use [`helm-docs`](https://github.com/norwoodj/helm-docs) to generate readme
- Translating an internal chart we've been installing for our use to trial the router and related to #786 to have a public helm chart published

_n.b. I did the helm-docs validation with github actions as that's our internal setup, but just read the contributor guidelines about circle CI. I'll have to modify the PR validation to suit here_

### CODE REVIEW

**Comment by garypen** [comment]
Hi @damienpontifex . Thanks for the contribution! I'll work with you on this PR to get it merged.

File: chart/router/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      annotations:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      labels:
+        {{- include "router.selectorLabels" . | nindent 8 }}
+    spec:
+      {{- with .Values.imagePullSecrets }}
+      imagePullSecrets:
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
```
Comment (garypen) [inline chart/router/templates/deployment.yaml:26]: Service a/c to use:
```suggestion
      {{- end }}
      serviceAccountName: {{ include "router.serviceAccountName" . }}
```

File: chart/router/templates/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      containers:
+        - name: {{ .Chart.Name }}
+          securityContext:
+            {{- toYaml .Values.securityContext | nindent 12 }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          args:
+            - --hot-reload
+            - --config
+            - /app/configuration.yaml
```
Comment (garypen) [inline chart/router/templates/deployment.yaml:38]: I'm not sure we should hard-code those directly into the template. We can probably combine the command line args with the `configuration` section you added into `values.yaml`. How about a values.yaml that looks something like: 
```
router:
  configuration:
    server:
      listen: 0.0.0.0:80
  args:
    - --hot-reload
    - <etc...>
```
?

File: chart/router/templates/deployment.yaml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            httpGet:
+              path: /.well-known/apollo/server-health
+              port: http
+          resources:
+            {{- toYaml .Values.resources | nindent 12 }}
+          volumeMounts:
+            - name: router-configuration
+              mountPath: /app/configuration.yaml
+              subPath: configuration.yaml
+              readOnly: true
```
Comment (garypen) [inline chart/router/templates/deployment.yaml:69]: This needs to be configurable.

File: chart/router/templates/ingress.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+{{- if .Values.ingress.enabled -}}
+{{- $fullName := include "router.fullname" . -}}
+{{- $svcPort := .Values.service.port -}}
+{{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
+  {{- if not (hasKey .Values.ingress.annotations "kubernetes.io/ingress.class") }}
+  {{- $_ := set .Values.ingress.annotations "kubernetes.io/ingress.class" .Values.ingress.className}}
+  {{- end }}
+{{- end }}
+apiVersion: networking.k8s.io/v1
```
Comment (garypen) [inline chart/router/templates/ingress.yaml:9]: I think most people are on modern enough kubernetes now, but helm client 3.8.1 still does this:
```suggestion
{{- if semverCompare ">=1.19-0" .Capabilities.KubeVersion.GitVersion -}}
  apiVersion: networking.k8s.io/v1
{{- else if semverCompare ">=1.14-0" .Capabilities.KubeVersion.GitVersion -}}
apiVersion: networking.k8s.io/v1beta1
{{- else -}}
apiVersion: extensions/v1beta1
{{- end }}
```

File: chart/router/values.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+image:
+  repository: ghcr.io/apollographql/router
+  pullPolicy: IfNotPresent
+  # Overrides the image tag whose default is the chart appVersion.
+  tag: ""
+
+imagePullSecrets: []
+nameOverride: ""
+fullnameOverride: ""
+
```
Comment (garypen) [inline chart/router/values.yaml:23]: service a/c stuff:
```suggestion

serviceAccount:
  # Specifies whether a service account should be created
  create: true
  # Annotations to add to the service account
  annotations: {}
  # The name of the service account to use.
  # If not set and create is true, a name is generated using the fullname template
  name: ""
```

**Review by garypen — CHANGES_REQUESTED** [review]
This looks great! One general comment:

Could you move `<router>/chart` into: `<router>/helm/chart` ?
(we'd like to isolate helm stuff into its own sub-directory)

Mostly I think the rest of my comments can be summarised as:
 - please add back the service a/c stuff. That's generally useful and easy to disable.
 - consider introducing a `router` section for configuration of CLI etc...
 - keep some helm boilerplate from `helm create`

File: chart/router/Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,6 @@
+apiVersion: v2
+name: router
+description: Rust Graph Routing runtime for Apollo Federation
+type: application
+version: 0.1.0
+appVersion: "v0.1.0-preview.2"
```
[the remark this answers is not in the document]
Comment (damienpontifex) [reply chart/router/Chart.yaml]: Restored these comments

File: chart/router/templates/_helpers.tpl [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+{{- end }}
+
+{{/*
+Selector labels
+*/}}
+{{- define "router.selectorLabels" -}}
+app.kubernetes.io/name: {{ include "router.name" . }}
+app.kubernetes.io/instance: {{ .Release.Name }}
+{{- end }}
+
```
[the remark this answers is not in the document]
Comment (damienpontifex) [reply chart/router/templates/_helpers.tpl]: Restored the service account defaults that come with `helm create`

File: chart/router/templates/deployment.yaml (hunk 2 above)
Comment (damienpontifex) [reply chart/router/templates/deployment.yaml]: I've updated to this, but still open to feedback on usability

File: chart/router/templates/deployment.yaml (hunk 3 above)
Comment (damienpontifex) [reply chart/router/templates/deployment.yaml]: Do you mean configurable for the mount path? Or other configurable options?

File: chart/router/templates/ingress.yaml (hunk 4 above)
Comment (damienpontifex) [reply chart/router/templates/ingress.yaml]: Removed all these conditionals. I think it's compatible with kubernetes >= 1.19 overall now

File: chart/router/templates/deployment.yaml (hunk 3 above)
Comment (garypen) [reply chart/router/templates/deployment.yaml]: I don't think we should assume that people always want to mount a configuration. I haven't honestly thought through the details, but it feels like we only want to do this if we are hot reloading or specifying a custom configuration so maybe we could tie this all together somehow? I'm very open to suggestions here...

Comment (garypen) [reply chart/router/templates/deployment.yaml]: I'll leave this as is for now and re-visit in follow up work.

File: .github/workflows/helm-chart-ci.yaml [hunk 8]
```diff
@@ -0,0 +1,29 @@
+name: Helm chart CI
```
[the remark this answers is not in the document]
Comment (garypen) [reply .github/workflows/helm-chart-ci.yaml:1]: I'll remove this for now and we'll decide how to sort out Circle-CI later.

**Review by garypen — APPROVED** [review]

**Comment by damienpontifex** [comment]
Thanks for working on this @garypen , sorry some delays on my side. I'm happy to continue testing/trialling things as they progress in the repo here

**Pull request merged**